### ISSUE TITLE
multiply_tensors

### ISSUE THREAD

**Issue by benz871** [obryanlouis/qa#10]
How it the multiply_tensors function in tf_utils is different from the normal matrix multiply function? Can you explain with an example. Also did you implement the character level embedding as described in the original paper. Is the cuda_lstm that you declared is different from the one present in the tensorflow library?
How did you create the answers span in the model?

**Comment by obryanlouis** [comment]
Hi Benz871 - 
The multiply_tensors function is just a wrapper around tf.matmul. It will take the two tenors and multiply them as if the last dimension of the first and first dimension of the last should be multiplied. It does not take batches into account. For example, you would get these dimensional results
multiply_tenors([dim_a, dim_b, dim_c], [dim_c, dim_d, dim_e]) = [dim_a, dim_b, dim_d, dim_e]
multiply_tenors([batch_size, dim_a, dim_b], [dim_b, dim_c]) = [batch_size, dim_a, dim_c]
multiply_tenors([dim_a, dim_b], [dim_b]) = [dim_a]
and so on.
tf.matmul on the other hand requires the inner dimensions to be the same. It also supports batch multiplications such as
tf.matmul([batch_size, dim_a, dim_b], [batch_size, dim_b, dim_c]) = [batch_size, dim_a, dim_c]
but won't support something like the above:
tf.matmul([batch_size, dim_a, dim_b], [dim_b, dim_c]) <-- error!
I made the convenience function multiply_tensors mainly to be able to do the above.

I used a character embedding and then ran a bi-rnn over the  characters to get the hidden outputs; the final start/end hidden outputs are used for each word.
https://github.com/obryanlouis/qa/blob/master/model/input_util.py#L66

I made the wrapper for the cudnn lstm because it was easier at the time. But it is the same as the one in tensorflow. The newest version of tensorflow doesn't require you to keep track of a buffer for the cudnn lstm, so it is easier to use.

Do you mean the answer spans as the input data to the model or the model outputs? You can take a look at the preprocessing folder to see how I created them, and there are a few different model variations I implemented for getting the outputs. The one that worked the best for me was the "stochastic answer pointer"
https://github.com/obryanlouis/qa/blob/master/model/stochastic_answer_pointer.py
You can find the paper for it here: https://arxiv.org/pdf/1712.03556.pdf

**Comment by benz871** [comment]
About the answers I meant how you created spans for the input. Because some of the answers are not directly in the assage but rather they are a part of the words in the passage. Like six this the actual word in the passage and six is the answer .  So how did you created the answer span input. Did you just ignore those questions or did something else like finding the word that has the closest embedding to the answer in the context?

**Comment by obryanlouis** [comment]
Ah - yeah. So for that I just took the smallest surrounding subspan based on the indices provided in the training data. But it didn't happen that often (like a handful of samples if I recall). I don't think it would matter much no matter what you do with them.

**Comment by benz871** [comment]
When we are sending the answers in the model during the training process we are sending the start and end from the tokens after splitting up each sentence or from the original context that we read from the file? Because you said some answers were not in the context or they were present in some different form than the original answers that required extra logic to be figured out? What is meant by exact match in the accuracy, is it like the answer span produced by the model should match the answer span provided or the answer should be same as the original answer. Like I asked earlier if six is the answer and in the passage its sixth so the start should be from there(index of sixth) so if the model predicts sixth is it correct or wrong.
Is the answer span created after you tokenize the words or before it?

**Comment by obryanlouis** [comment]
The scoring is based on the original answer, so if the training data had "sixth" and the right answer was "six" then it would be wrong. It is intentionally strict, but that case doesn't happen often.
